### PR TITLE
Points calculation

### PR DESCRIPTION
**Pull request by netum-heikki**

### CODE REVIEW

File: src/app/backing-up-from-results-not-allowed-guard.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import { ActivatedRouteSnapshot, CanActivate, RouterStateSnapshot } from '@angular/router';
+@Injectable()
+export class BackingUpFromResultsNotAllowedGuard implements CanActivate {
+
+  previousPageResultsOrUndefined = true;
+
+  constructor() { }
+
+  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): boolean {
+    const access = !this.previousPageResultsOrUndefined || route.url.toString().indexOf('kysymykset') === -1;
```
Comment (hollannikas) [inline src/app/backing-up-from-results-not-allowed-guard.ts]: Can you not handle this with _only_ the provided router snapshots? That way canActivate is a pure function.

File: src/app/backing-up-from-results-not-allowed-guard.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+import { ActivatedRouteSnapshot, CanActivate, RouterStateSnapshot } from '@angular/router';
+@Injectable()
+export class BackingUpFromResultsNotAllowedGuard implements CanActivate {
```
Comment (hollannikas) [inline src/app/backing-up-from-results-not-allowed-guard.ts]: As much as I am all for explicit naming, this is a bit on the long side.

**Review by hollannikas — CHANGES_REQUESTED** [review]

**Comment by hollannikas** [comment]
It might be good to change my "Ammend me" commit to a more descriptive comment. I left it like that as a reminder for myself to ammed that commit (and change the comment) 😉

File: src/app/questions/questions.component.ts [hunk 3]
```diff
@@ -20,11 +18,15 @@ import { Subscription } from 'rxjs/Subscription';
...
-export class QuestionsComponent implements OnInit {
+export class QuestionsComponent implements OnInit, OnDestroy {
 
   questions: Observable<Question[]>;
   question: Observable<Question>;
 
+  answersOfSelectedQuestion: Observable<Answer[]>;
+
+  subscriptions: Subscription[] = [];
+
```
Comment (hollannikas) [inline src/app/questions/questions.component.ts]: This seems to always be an array of one 😀  Please change this to `subscription: Subscription`

File: src/app/reducers/questions.ts [hunk 4]
```diff
@@ -49,12 +53,18 @@ export function reducer(state = initialState, action: questions.Actions): State
       const points = calculateUserPoints(state.pickedQuestions[questionIndex], action.answers);
       const newPointsState = [...state.questionPoints];
       newPointsState[questionIndex] = points;
+      // all previous answers filtered out from new state and new answers added
+      const newAllAnswersState = state.allAnswers
+        .filter(existingAnswer => !action.answers.some(newAnswer => existingAnswer.optionId === newAnswer.optionId));
+      action.answers.forEach(replacingAnswer => newAllAnswersState.push(replacingAnswer));
+
       const newState = {
```
Comment (hollannikas) [inline src/app/reducers/questions.ts:61]: This is a bit overcomplicated for a simple replace. Why not group the answers per question, for example `{ questionId: Answer[] }` or something? That way you can use something like `allAnswers[questionId] = replacingAnswers`

**Review by hollannikas — CHANGES_REQUESTED** [review]

File: src/app/questions/questions.component.ts (hunk 3 above)
Comment (netum-heikki) [reply src/app/questions/questions.component.ts]: I'll change it. I just like it because it takes care of all subscriptions regardless of the amount.

**Pull request merged**